File: src/main/java/com/somefrills/features/farming/autofarmer/RectangularMovement.java
package com.somefrills.features.farming.autofarmer;

/**
 * Rectangular pattern farming (melon, pumpkin, wheat, carrot, potato, nether wart).
 * Cycles through: Forward (with forward always pressed) → Left (with forward always pressed) → repeat
 */
public class RectangularMovement implements MovementStrategy {
    private State currentState = State.RIGHT;

    public RectangularMovement() {
    }

    public MovementState getCurrentState() {
        return currentState.toMovementState();
    }

    public void nextState() {
        currentState = currentState.next();
    }

    private enum State {
        RIGHT {
            @Override
            public State next() {
                return LEFT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, false, true);
            }
        },
        LEFT {
            @Override
            public State next() {
                return RIGHT;
            }

            @Override
            public MovementState toMovementState() {
                return new MovementState(true, false, true, false);
            }
        };

        protected abstract State next();

        protected abstract MovementState toMovementState();
    }
}

File: src/main/kotlin/com/somefrills/features/farming/autofarmer/RectangularMovement.kt
package com.somefrills.features.farming.autofarmer

/**
 * Rectangular pattern farming (melon, pumpkin, wheat, carrot, potato, nether wart).
 * Cycles through: Forward+Right → Forward+Left → repeat
 */
class RectangularMovement : MovementStrategy() {

    private var state: FarmingState = FarmingState.RIGHT

    override fun getState(): MovementState {
        return state.toMovementState()
    }

    override fun nextState() {
        state = state.next() as FarmingState
    }

    private enum class FarmingState : State {
        RIGHT {
            override fun next() = LEFT

            override fun toMovementState(): MovementState {
                return MovementState(
                    MovementState.FORWARD or MovementState.RIGHT
                )
            }
        },

        LEFT {
            override fun next() = RIGHT

            override fun toMovementState(): MovementState {
                return MovementState(
                    MovementState.FORWARD or MovementState.LEFT
                )
            }
        };
    }
}